File: structural-analysis/src/aec_structural_validation/response_spectrum.py
"""Response-spectrum post-processing helpers validated against SEIS-01.

Only global base-shear scaling and orthogonal SRSS combination are included.
No eigensolution, modal extraction, spectrum interpolation, within-direction
CQC, or full dynamic solver is implemented here.
"""
from __future__ import annotations

import math
from typing import Iterable, Mapping


def global_scale_factor(design_base_shear: float, minimum_dynamic_base_shear: float) -> float:
    if minimum_dynamic_base_shear <= 0:
        raise ValueError("minimum_dynamic_base_shear must be positive")
    return float(design_base_shear) / float(minimum_dynamic_base_shear)


def scale_response(response: Mapping[str, float], factor: float) -> dict[str, float]:
    return {key: float(value) * float(factor) for key, value in response.items()}


def srss(components: Iterable[float]) -> float:
    return math.sqrt(sum(float(value) ** 2 for value in components))


def principal_direction_scaling(inputs: Mapping[str, object]) -> dict[str, object]:
    factor = global_scale_factor(float(inputs["design_base_shear_kips"]), float(inputs["minimum_dynamic_base_shear_kips"]))
    return {
        "scale_factor": factor,
        "scaled_major": scale_response(inputs["unscaled_major"], factor),
        "scaled_minor": scale_response(inputs["unscaled_minor"], factor),
    }

File: public/structural-demo/solver/response-spectrum.js
export function globalScaleFactor(designBaseShear, minimumDynamicBaseShear) {
  if (!(Number(minimumDynamicBaseShear) > 0)) throw new Error("minimumDynamicBaseShear must be positive");
  return Number(designBaseShear) / Number(minimumDynamicBaseShear);
}
export function scaleResponse(response, factor) {
  return Object.fromEntries(Object.entries(response).map(([key, value]) => [key, Number(value) * Number(factor)]));
}
export function srss(components) {
  return Math.sqrt([...components].reduce((sum, value) => sum + Number(value) ** 2, 0));
}
export function principalDirectionScaling(inputs) {
  const factor = globalScaleFactor(inputs.designBaseShearKips ?? inputs.design_base_shear_kips, inputs.minimumDynamicBaseShearKips ?? inputs.minimum_dynamic_base_shear_kips);
  return {
    scaleFactor: factor,
    scaledMajor: scaleResponse(inputs.unscaledMajor ?? inputs.unscaled_major, factor),
    scaledMinor: scaleResponse(inputs.unscaledMinor ?? inputs.unscaled_minor, factor),
  };
}
